### PR TITLE
feat: Update unevalTree when setter methods are executed and run evaluations for dependent properties and resolve promise

### LINKED ISSUE #22705: Phase 1 [SUBTASK] Run evaluation for the dependent property when a worker is idle and resolve the promise of the setter method.

### PR DESCRIPTION
**Pull request by Druthi**
## Description
1. Update unevalTree when setter methods are executed
2. Run evaluation for the dependent property when a worker is idle and resolve the promise of the setter method.

Fixes #22705 and #22707 

#### Type of change
- New feature (non-breaking change which adds functionality)
- This change requires a documentation update

- [ ] Manual
- [ ] Jest
- [ ] Cypress
>
>
#### Dev activity
- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] PR is being merged under a feature flag

#### QA activity:
- [ ] [Speedbreak features](https://github.com/appsmithorg/TestSmith/wiki/Test-plan-implementation#speedbreaker-features-to-consider-for-every-change) have been covered
- [ ] Test plan covers all impacted features and [areas of interest](https://github.com/appsmithorg/TestSmith/wiki/Guidelines-for-test-plans/_edit#areas-of-interest)
- [ ] Test plan has been peer reviewed by project stakeholders and other QA members
- [ ] Manually tested functionality on DP
- [ ] We had an implementation alignment call with stakeholders post QA Round 2
- [ ] Cypress test cases have been added and approved by SDET/manual QA
- [ ] Added `Test Plan Approved` label after Cypress tests were reviewed
- [ ] Added `Test Plan Approved` label after JUnit tests were reviewed

### CODE REVIEW

File: app/client/src/sagas/EvaluationsSaga.ts [hunk 1]
```diff
@@ -139,7 +140,7 @@ export function* updateDataTreeHandler(
     errors,
     evalMetaUpdates = [],
     evaluationOrder,
-    jsUpdates,
+    jsUpdates = {},
```
Comment (Druthi) [inline app/client/src/sagas/EvaluationsSaga.ts:143]: Temporary change

File: app/client/src/workers/Evaluation/evaluate.ts [hunk 2]
```diff
@@ -225,7 +228,30 @@ export function setEvalContext({
     isTriggerBased,
   });
 
-  Object.assign(self, evalContext);
+  Object.assign(self, klona(evalContext));
```
Comment (Druthi) [inline app/client/src/workers/Evaluation/evaluate.ts:231]: Temp changes

File: app/client/src/workers/Evaluation/evalTreeWithChanges.ts [hunk 3]
```diff
@@ -14,7 +14,10 @@ import { MAIN_THREAD_ACTION } from "@appsmith/workers/Evaluation/evalWorkerActio
 import type { UpdateDataTreeMessageData } from "sagas/EvalWorkerActionSagas";
 import type { JSUpdate } from "utils/JSPaneUtils";
 
-export function evalTreeWithChanges(updatedValuePaths: string[][]) {
+export function evalTreeWithChanges(
+  updatedValuePaths: string[][],
+  callback?: any,
```
Comment (Rishabh-Rathod) [inline app/client/src/workers/Evaluation/evalTreeWithChanges.ts:19]: Let's avoid using `any` for type wherever possible.

**Review by Rishabh-Rathod — CHANGES_REQUESTED** [review]

File: app/client/src/ce/workers/Evaluation/Actions.ts [hunk 4]
```diff
@@ -37,12 +38,17 @@ export enum ExecutionType {
 
 function getEntityMethodFromConfig(entityConfig: DataTreeEntityConfig) {
   const setterMethodMap: Record<string, any> = {};
-  // @ts-expect-error: a
+
+  if (!isWidget(entityConfig)) return setterMethodMap;
+
   if (entityConfig.__setters) {
```
[the remark this answers is not in the document]
Comment (Druthi) [reply app/client/src/ce/workers/Evaluation/Actions.ts:44]: But typescript throws errors if we dont

File: app/client/src/workers/common/DataTreeEvaluator/index.ts [hunk 5]
```diff
@@ -1660,6 +1666,58 @@ export default class DataTreeEvaluator {
...
+      propertyPath,
+      value,
+      currentTree: this.evalTree,
+      configTree: this.oldConfigTree,
+      evalMetaUpdates,
+      fullPropertyPath: path,
+      isNewWidget: false,
+      shouldUpdateGlobalContext: true,
+      overriddenProperties,
+    });
```
[the remark this answers is not in the document]
Comment (Druthi) [reply app/client/src/workers/common/DataTreeEvaluator/index.ts:1690]: But we need  `overriddenProperties` for `updatedProperties` to be passed to `evalTreeWithChanges`

File: app/client/src/workers/Evaluation/evalTreeWithChanges.ts [hunk 6]
```diff
@@ -81,4 +87,16 @@ export function evalTreeWithChanges(updatedValuePaths: string[][]) {
...
     },
   });
+
+  if (callback) {
+    if (shouldResetEvalContext) {
+      setEvalContext({
+        dataTree,
+        isDataField: false,
+        isTriggerBased: true,
+      });
```
Comment (Rishabh-Rathod) [inline app/client/src/workers/Evaluation/evalTreeWithChanges.ts:97]: This should happen every time after the sync evaluation, It was a miss in JSObject variable mutation feature. I will create a PR to fix this.

File: app/client/src/workers/common/DataTreeEvaluator/index.ts [hunk 7]
```diff
@@ -1660,6 +1666,58 @@ export default class DataTreeEvaluator {
...
+        overriddenProperties,
+      });
+
+      overriddenProperties.forEach((propPath) => {
+        updatedProperties.push([entityName, propPath]);
+
+        if (propPath.split(".")[0] === "meta") {
+          const metaPropertyPath = propPath.split(".").slice(1);
+
+          evalMetaUpdates.push({
```
Comment (Rishabh-Rathod) [inline app/client/src/workers/common/DataTreeEvaluator/index.ts:1697]: let's move this method to separate file and use setEvalTree to update evalTree

File: app/client/src/workers/Evaluation/evalTreeWithChanges.ts (hunk 6 above)
Comment (Rishabh-Rathod) [inline app/client/src/workers/Evaluation/evalTreeWithChanges.ts:97]: This was supposed to be there for all the cases, it was missed in JSObject variable mutation feature. I will create a separate PR to add this.

Comment (Druthi) [reply app/client/src/workers/Evaluation/evalTreeWithChanges.ts:97]: Cool makes sense. Shall I remove it from here then?

File: app/client/src/workers/Evaluation/setters.ts [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+          data: { evalMetaUpdates },
+        });
+      }
+    });
+  }
+
+  set(evalTree, path, value);
+  set(self, path, value);
+
+  // dataTreeEvaluator?.setEvalTree(evalTree);
```
Comment (Rishabh-Rathod) [inline app/client/src/workers/Evaluation/setters.ts:62]: nit: remove the comment

Comment (Druthi) [reply app/client/src/workers/Evaluation/setters.ts]: Done

**Review by Rishabh-Rathod — APPROVED** [review]

**Pull request merged**